[Merged by Bors] - feat(analysis/normed/group/add_circle): results about finite-order points on the circle

ocfnash opened the pull request:
The main motivation for developing this API is the lemma `add_circle.le_add_order_smul_norm_of_is_of_fin_add_order`

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(analysis/normed/group/add_circle): results about finite-order points on the circle

ocfnash committed:
Merge branch 'master' into ocfnash/circle_finite_order

Hoping to fix a deterministic timeout unrelated to my work

ocfnash committed:
Fix lint

ocfnash committed:
Merge branch 'master' into ocfnash/circle_finite_order

j-loreaux commented on src/topology/instances/add_circle.lean:
```diff
@@ -54,6 +58,26 @@ section linear_ordered_field
 
+lemma coe_nsmul {n : ℕ} {x : 𝕜} : (↑(n • x) : add_circle p) = n • (x : add_circle p) := rfl
+
+lemma coe_zsmul {n : ℤ} {x : 𝕜} : (↑(n • x) : add_circle p) = n • (x : add_circle p) := rfl
```
Can you take advantage of #17048 now that it is merged? Otherwise these lemmas should be protected, but it would be cool to just not need them at all.
```suggestion
lemma coe_nsmul {n : ℕ} {x : 𝕜} : (↑(n • x) : add_circle p) = n • (x : add_circle p) := rfl

lemma coe_zsmul {n : ℤ} {x : 𝕜} : (↑(n • x) : add_circle p) = n • (x : add_circle p) := rfl
```

ocfnash replied:
Great idea, I can indeed. I have just dropped these and added `instance : coe_is_add_monoid_hom 𝕜 (add_circle p) :=` instead.

j-loreaux commented on src/topology/instances/add_circle.lean:
```diff
@@ -126,6 +154,86 @@ instance : divisible_by (add_circle p) ℤ :=
+
+/-- See also `add_circle.gcd_mul_add_order_of_div_eq`. -/
+lemma add_order_of_div_of_gcd_eq_one {m n : ℕ} (hn : 0 < n) (h : gcd n m = 1) :
+  add_order_of (↑(↑m / ↑n * p) : add_circle p) = n :=
```
Would it be useful to have a version of this for `ℚ` as well? Or is the problem that there is no coercion from `ℚ` to `add_circle p`?

ocfnash replied:
A version for `ℚ` is a good idea. I have just added the lemmas `add_order_of_div_of_gcd_eq_one'` and `add_order_of_coe_rat` to cater for this.

Since you mention it, there is no coercion from `ℚ` because it would only be useful if it also multiplied by the period `p` but then it would disagree with the real version (unless `p = 1`). So for now at least I'm just not defining the coercion.

j-loreaux commented on src/topology/instances/add_circle.lean:
```diff
@@ -126,6 +154,86 @@ instance : divisible_by (add_circle p) ℤ :=
+  add_order_of (↑(↑m / ↑n * p) : add_circle p) = n :=
+begin
+  rcases m.eq_zero_or_pos with rfl | hm, { rw [gcd_zero_right, normalize_eq] at h, simp [h], },
+  let x : add_circle p := ↑(↑m / ↑n * p),
```
not required, but I think this makes the goal more readable
```suggestion
  set x : add_circle p := ↑(↑m / ↑n * p),
```

ocfnash replied:
OK done.

j-loreaux commented on src/topology/instances/add_circle.lean:
```diff
@@ -126,6 +154,86 @@ instance : divisible_by (add_circle p) ℤ :=
+       ↑(((m : 𝕜) / add_order_of u) * p) = u :=
+begin
+  rcases eq_or_ne u 0 with rfl | hu, { exact ⟨0, by simp⟩, },
+  let n := add_order_of u,
```
same here
```suggestion
  set n := add_order_of u,
```

ocfnash replied:
OK done.

j-loreaux reviewed (commented):
This feels like it was more work than it should have been, but I don't see any obvious way to simplify in any significant way.

ocfnash committed:
Merge branch 'master' into ocfnash/circle_finite_order

ocfnash committed:
Merge branch 'master' into ocfnash/circle_finite_order

ocfnash committed:
Suggestions from CR

The number of lines changed is slightly higher than it might be because
I have also flipped the order of arguments in `gcd n m` to `gcd m n`.
This is to be compatible with `rat.cop` which became relevant in view of
the new lemma `add_circle.add_order_of_coe_rat`.

eric-wieser commented on src/algebra/order/floor.lean:
```diff
@@ -678,6 +678,19 @@ sub_nonneg_of_le $ (le_div_iff hb).1 $ floor_le _
 lemma sub_floor_div_mul_lt (a : k) (hb : 0 < b) : a - ⌊a / b⌋ * b < b :=
 sub_lt_iff_lt_add.2 $ by { rw [←one_add_mul, ←div_lt_iff hb, add_comm], exact lt_floor_add_one _ }
 
+lemma fract_div_coe_eq_div_coe_mod {m n : ℕ} :
```
This should probably have `nat_cast` instead of `coe`; is a similar lemma true for `int`?

ocfnash replied:
OK I've just added https://github.com/leanprover-community/mathlib/pull/17321/commits/dfa0ec223612857a34f38048e0a090186bab3a77 which attempts to address this comment.

Note that the new lemma caters for `{m : ℤ} {n : ℕ}` but not `{m n : ℤ}` since there is no clean result in this case (most easily seen by consulting `int.fract_neg`).

ocfnash committed:
Suggestions from CR:

 * Renaming: use `nat_cast` instead of `coe` in various places
 * Add lemma `fract_div_int_cast_eq_div_int_cast_mod` supporting `int`

eric-wieser commented on src/algebra/order/floor.lean:
```diff
@@ -678,6 +678,47 @@ sub_nonneg_of_le $ (le_div_iff hb).1 $ floor_le _
+    rw [← nat.cast_add, nat.mod_add_div m n], },
+end
+
+lemma fract_div_int_cast_eq_div_int_cast_mod {m : ℤ} {n : ℕ} :
+  fract ((m : k) / n) = ↑(m % n) / n :=
```
I think it is true for both arguments integers as
```suggestion
lemma fract_div_int_cast_eq_div_int_cast_fmod {m n : ℤ} :
  int.fract ((m : ℚ) / n) = ↑(m.fmod n) / n :=
```

ocfnash replied:
I'm very reluctant to do this because:
 * Mathlib contains literally zero API for `int.fmod`,
 * I'm already yak shaving and currently short of time.

As a compromise I have added a comment proposing the `int.fmod` version.

eric-wieser replied:
Thanks! Certainly doing that is out of scope for this PR, I should have been clearer about that. I thought you'd be interested to know the version which is true though.

eric-wieser commented on src/algebra/order/floor.lean:
```diff
@@ -678,6 +678,47 @@ sub_nonneg_of_le $ (le_div_iff hb).1 $ floor_le _
+    { rw [of_nat_eq_coe, cast_coe_nat, ← coe_nat_mod, cast_coe_nat,
+        fract_div_nat_cast_eq_div_nat_cast_mod], },
+    { rw [neg_succ_not_nonneg] at hl, contradiction, }, },
+  induction m, { exact this (of_nat_nonneg m), },
```
Does using `int.eq_coe_or_neg` make the proof any easier?

ocfnash replied:
Yes it does, thanks.

ocfnash committed:
Simplify proof slightly using `int.eq_coe_or_neg`.

eric-wieser commented on src/algebra/order/floor.lean:
```diff
@@ -678,6 +678,50 @@ sub_nonneg_of_le $ (le_div_iff hb).1 $ floor_le _
+    { rw [cast_coe_nat, ← coe_nat_mod, cast_coe_nat, fract_div_nat_cast_eq_div_nat_cast_mod], },
+    { rw [right.nonneg_neg_iff, coe_nat_nonpos_iff] at hl, simp [hl, zero_mod], }, },
+  obtain ⟨m₀, hm₀⟩ := m.eq_coe_or_neg,
+  rcases hm₀ with rfl | rfl, { exact this (of_nat_nonneg m₀), },
```
```suggestion
  obtain ⟨m₀, rfl | rfl⟩ := m.eq_coe_or_neg,
  { exact this (of_nat_nonneg m₀), },
```

ocfnash replied:
Thanks, done!

ocfnash committed:
Some golfing

eric-wieser committed:
Merge branch 'master' into ocfnash/circle_finite_order

eric-wieser commented on src/analysis/normed/group/add_circle.lean:
```diff
@@ -211,6 +218,43 @@ begin
+include hp
+
+lemma norm_div_nat_cast {m n : ℕ} :
+  ∥(↑((↑m / ↑n) * p) : add_circle p)∥ = p * (↑(min (m % n) (n - m % n)) / n) :=
```
I think you'll need to update the norm symbols to make this pass CI

ocfnash replied:
Thanks, done.

ocfnash committed:
Fix after merge master
